### PR TITLE
feat(ecocredit/core): dynamic batch minting

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

- implements dynamic batch minting server impl
- refactors calls to fetch credit balances into a util that handles `Not Found` cases.
- implements tests for the added util and the existing util functions

ref: #973 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/server/core/mint_batch_credits.go [hunk 1]
```diff
@@ -3,9 +3,139 @@ package core
...
+			if err != nil {
+				return nil, err
+			}
+			if err := sdkCtx.EventManager().EmitTypedEvent(&core.EventReceive{
+				Sender:         req.Issuer,
+				Recipient:      iss.Recipient,
+				BatchDenom:     req.BatchDenom,
+				TradableAmount: iss.TradableAmount,
+				RetiredAmount:  iss.RetiredAmount,
+				BasketDenom:    "",
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits.go:107]:
```suggestion
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	assert.NilError(t, err)
+	supplyBefore, err := s.stateStore.BatchSupplyTable().Get(ctx, batch.Key)
+	assert.NilError(t, err)
+
+	mintTradable, mintRetired := math.NewDecFromInt64(10), math.NewDecFromInt64(10)
+	msg := core.MsgMintBatchCredits{
+		Issuer:     s.addr.String(),
+		BatchDenom: batch.Denom,
+		Issuance: []*core.BatchIssuance{
+			{Recipient: s.addr.String(), TradableAmount: mintTradable.String(), RetiredAmount: mintRetired.String(), RetirementJurisdiction: "US-OR"},
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:33]:
```suggestion
			{
				Recipient: s.addr.String(),
				TradableAmount: mintTradable.String(),
				RetiredAmount: mintRetired.String(),
				RetirementJurisdiction: "US-OR",
			},
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+	_, err = s.k.MintBatchCredits(ctx, &msg)
+	assert.NilError(t, err)
+
+	balAfter, err := s.stateStore.BatchBalanceTable().Get(ctx, s.addr, batch.Key)
+	assert.NilError(t, err)
+	supplyAfter, err := s.stateStore.BatchSupplyTable().Get(ctx, batch.Key)
+	assert.NilError(t, err)
+
+	assertCreditsMinted(t, balBefore, balAfter, supplyBefore, supplyAfter, mintTradable, mintRetired, math.NewDecFromInt64(0), 6)
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:50]: Might be nice to use a struct given there are nine parameters here.

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+func TestMintBatchCredits_MintToNewAccount(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	ctx := s.ctx
+	batch := setupMintBatchTest(s, true)
+
+	newAcc := sdk.AccAddress("NewAccount")
+	balBefore, err := s.stateStore.BatchBalanceTable().Get(ctx, newAcc, batch.Key)
+	assert.ErrorContains(t, err, ormerrors.NotFound.Error())
+	balBefore = &api.BatchBalance{BatchKey: batch.Key, Address: newAcc}
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:63]: No sure if it's necessary to test balance not found given we create the account in the line above.

```suggestion
	newAcc := sdk.AccAddress("NewAccount")
	balBefore := &api.BatchBalance{BatchKey: batch.Key, Address: newAcc}
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	supplyBefore, err := s.stateStore.BatchSupplyTable().Get(ctx, batch.Key)
+	assert.NilError(t, err)
+
+	mintTradable, mintRetired := math.NewDecFromInt64(10), math.NewDecFromInt64(10)
+	msg := core.MsgMintBatchCredits{
+		Issuer:     s.addr.String(),
+		BatchDenom: batch.Denom,
+		Issuance: []*core.BatchIssuance{
+			{Recipient: newAcc.String(), TradableAmount: mintTradable.String(), RetiredAmount: mintRetired.String(), RetirementJurisdiction: "US-OR"},
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:72]:
```suggestion
			{
				Recipient: newAcc.String(),
				TradableAmount: mintTradable.String(),
				RetiredAmount: mintRetired.String(),
				RetirementJurisdiction: "US-OR",
			},
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	balAfter, err := s.stateStore.BatchBalanceTable().Get(ctx, s.addr, batch.Key)
+	assert.NilError(t, err)
+	supplyAfter, err := s.stateStore.BatchSupplyTable().Get(ctx, batch.Key)
+	assert.NilError(t, err)
+
+	assertCreditsMinted(t, balBefore, balAfter, supplyBefore, supplyAfter, mintTradable, mintRetired, math.NewDecFromInt64(0), 6)
+}
+
+// tests that batch credits can be minted and sent to an account with no previous balance
+func TestMintBatchCredits_MintToNewAccount(t *testing.T) {
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:54]: I'm not sure if we need a test for minting to an account with an empty balance as opposed to an account with a positive balance. Curious if removing this test would change the code coverage at all as I think what is really being tested here is the `GetBalance` method already tested in `TestUtils_GetBalance`.

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	s := setupBase(t)
+	batch := setupMintBatchTest(s, true)
+	addr := sdk.AccAddress("foobar")
+
+	_, err := s.k.MintBatchCredits(s.ctx, &core.MsgMintBatchCredits{
+		Issuer:     addr.String(),
+		BatchDenom: batch.Denom,
+		Issuance:   nil,
+		OriginTx:   nil,
+		Note:       "",
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:103]:
```suggestion
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	s := setupBase(t)
+	batch := setupMintBatchTest(s, false)
+	addr := sdk.AccAddress("foobar")
+
+	_, err := s.k.MintBatchCredits(s.ctx, &core.MsgMintBatchCredits{
+		Issuer:     addr.String(),
+		BatchDenom: batch.Denom,
+		Issuance:   nil,
+		OriginTx:   nil,
+		Note:       "",
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:119]:
```suggestion
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+	s := setupBase(t)
+	setupMintBatchTest(s, true)
+	addr := sdk.AccAddress("foobar")
+
+	_, err := s.k.MintBatchCredits(s.ctx, &core.MsgMintBatchCredits{
+		Issuer:     addr.String(),
+		BatchDenom: "C05-00000000-00000000-001",
+		Issuance:   nil,
+		OriginTx:   nil,
+		Note:       "",
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:135]:
```suggestion
```

File: x/ecocredit/server/core/mint_batch_credits_test.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+	ctx := s.ctx
+	_, _, batchDenom := s.setupClassProjectBatch(s.t)
+	batch, err := s.stateStore.BatchTable().GetByDenom(ctx, batchDenom)
+	assert.NilError(s.t, err)
+	batch.Open = open
+	assert.NilError(s.t, s.stateStore.BatchTable().Update(ctx, batch))
+	return batch
+}
+
+func assertCreditsMinted(t *testing.T, balBefore, balAfter *api.BatchBalance, supBefore, supAfter *api.BatchSupply, tradable, retired, escrowed math.Dec, precision uint32) {
```
Comment (ryanchristo) [inline x/ecocredit/server/core/mint_batch_credits_test.go:150]: As mentioned above, a struct might be nice given there are nine parameters here.

File: x/ecocredit/server/utils/utils_test.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	assert.NilError(t, s.stateStore.ClassTable().Insert(s.ctx, &api.Class{
+		Id:               "C01",
+		Admin:            s.addr,
+		Metadata:         "foo",
+		CreditTypeAbbrev: "C",
+	}))
+	batchDenom := "C01-000000-0000000-001"
+	ct, err := GetCreditTypeFromBatchDenom(s.ctx, s.stateStore, batchDenom)
+	assert.NilError(t, err)
+	assert.DeepEqual(t, ct, creditType, cmpopts.IgnoreUnexported(api.CreditType{}))
```
Comment (ryanchristo) [inline x/ecocredit/server/utils/utils_test.go:113]: Is ignore unexported necessary here?

**Review by ryanchristo — COMMENTED** [review]
Looking good. A few questions and some minor suggestions.

File: x/ecocredit/server/utils/utils_test.go (hunk 11 above)
Comment (technicallyty) [reply x/ecocredit/server/utils/utils_test.go:113]: yes, Pulsar types have private fields that should not be compared. an error is returned without it due to the comparison of the `state` field

File: x/ecocredit/server/core/mint_batch_credits_test.go (hunk 10 above)
Comment (technicallyty) [reply x/ecocredit/server/core/mint_batch_credits_test.go]: reduced it a bit here - https://github.com/regen-network/regen-ledger/pull/1059/commits/5268b06df0a06a0ce8a203e65b8fa2babec65c5b

passing the issuance as a whole rather than individual decimals. lmk if you think this should be further reduced

File: x/ecocredit/server/core/mint_batch_credits_test.go (hunk 6 above)
Comment (technicallyty) [reply x/ecocredit/server/core/mint_batch_credits_test.go]: makes sense, removed the test. all the other tests test against existing balance set up in `setupClassProjectBatch`

File: x/ecocredit/server/core/mint_batch_credits_test.go (hunk 10 above)
Comment (ryanchristo) [reply x/ecocredit/server/core/mint_batch_credits_test.go]: That looks good to me.

**Review by ryanchristo — APPROVED** [review]
Looks good to me.

**Review by aleem1314 — APPROVED** [review]
lgtm

**Comment by ryanchristo** [comment]
@technicallyty looks like the merge caused duplicate imports. Can you fix? :pray:
